// omni/index.js
var through = require('through2') // fix this thing up!

module.exports = function Actions (C) {
  var omnibox
  var menu
  var search

  var s = through.obj(function Write (chunk, enc, next) { 
    console.log(chunk)
    next()
  })

  if (localStorage.library && localStorage.library.length > 0) {
    var library = JSON.parse(localStorage.library)
  } else library = null // request library

  window.addEventListener('contextmenu', ctx, false)
  window.addEventListener('keydown', function toggleOmni (e) {
    if (e.keyCode === 9) {
      e.preventDefault();
      (C.live === false) ? C.draw() : C.erase()
      return false
    }
  }, false) 

  function ctx (e) {
    e.preventDefault()
    var element = e.target
    var id = element.id
    var parent = element.parentElement // ?
    var cl = element.className
    var tag = element.tagName
    console.log(element, parent, parent.id, tag, cl)
  }

  function updateSearch (e) {
    e.preventDefault()
    omnibox.innerHTML = ''
    var str = search.value
    var res = find(library, str) 

    if (res) 
      omnibox.innerHTML = JSON.stringify(res) // package selected

    switch (e.keyCode) {
      case 38 : ; break; // up
      case 40 : console.log(res.name, res); break; // act on selection!
      case 13 : console.log(res.name, res); break; // act on selection!
      default: console.log(e);
    }
  }

  function omniNav (e) {
    if (e.keyCode === 40) console.log('toggle to input')

  }

  C.unbind = function () {
    search.removeEventListener('keyup', updateSearch)
  }

  C.bind = function (domNode) {
    omnibox = domNode.querySelector('.omnibox')
    search = domNode.querySelector('#search')
    search.addEventListener('keyup', updateSearch, false)
    search.focus()
  }

  return s
}

function find (haystack, needle) {
  for (hay in haystack) {
    if (needle !== '' && hay.match(needle)) return haystack[hay]
  }
}
