// omni/index.js
var D = require('browser/mini') // domrender
var asyncMap = require('slide').asyncMap
var through = require('through2')
var utils = require('utils')
var cancel = utils.cancel


module.exports = function Actions (id, template) {

  var library = (!localStorage.library) 
    ? null // get data from db
    : JSON.parse(localStorage.library)

  var s = through.obj(function Write (d, enc, next) { 
    if (d.to && d.to === id) console.log('omni', d)
    next()
  })

  var filters = [['library', 'put', 'delete'],
                 ['layers', 'open', 'delete'],
                 ['canvas', 'save', 'clear']]

  var mode = 'filter'
  var selFilter = 0
  var selAction = 0
  var omniNode
  var buffer
  var live

  // dom >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  var om = D(document.body, template.get('.omni'), utils.UID('omni'))
  var disp
  var act

  om(function (el) {
    omniNode = el
    live = true
    disp = D(el.querySelector('.omnibox'), template.get('.res'))
    act = D(el.querySelector('.actions'), template.get('.label'))
    fil = D(el.querySelector('.filter'), template.get('.label'))
    fil({txt:filters[selFilter][0]})
    el.querySelector('#search').addEventListener('keyup', keyInput, false)
    el.querySelector('#search').focus()
  }) // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<


  // key interactions >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  window.addEventListener('keydown', keyMappings, false)
  window.addEventListener('contextmenu', contextMenu, false)

  function keyInput (e) { // execute search based on current filter
    e.preventDefault()
    var str = e.target.value
    var res = utils.search(library.master, str)

    if (res && str != '') {
      buffer = res
      disp({txt:res.name})
      act({txt:'> '+filters[selFilter].slice(1)[selAction]})
    } else if (!res || str === '') {
      buffer = null
      disp({txt:''})
      act({txt:''})
    }
  }

  function toggle () {
    if (live === true) {
      live = false
      omniNode.style.display = 'none'
    } else if (live === false) {
      live = true
      omniNode.style.display = 'block'
      document.body.querySelector('#search').focus()
    }
  }

  function sel (incdec) {
    var actions = filters[selFilter].slice(1)

    if (mode === 'filter') {
      selFilter = incDec(incdec, selFilter, 0, filters.length-1)
      fil({txt:filters[selFilter][0]})
    } else if (mode === 'action') {
      selAction = incDec(incdec, selAction, 0, actions.length-1)
      act({txt:'> '+actions[selAction]})
    }

    function incDec (incdec, val, min, max) {
      if (incdec==='inc') val++
      if (incdec==='dec') val--
      if (val < min) val = min
      if (val > max) val = max
      return val
    }
  }

  function exec () {
    console.log('bam')
    var action = filters[selFilter].slice(1)[selAction]
    s.push({
      key : action,
      value : buffer.name
    })
  }

  function toggleMode () {
    if (mode === 'filter') mode = 'action'
    else if (mode === 'action') mode = 'filter'
  }

  function keyMappings (e) {
    if (e.keyCode === 27) { e.preventDefault(); toggle() }
    if (e.keyCode === 38 && live === true) { e.preventDefault(); sel('inc') }
    if (e.keyCode === 40 && live === true) { e.preventDefault(); sel('dec') }
    if (e.keyCode === 9 && live === true) { e.preventDefault(); toggleMode() }
    if (e.keyCode === 13 && live === true) { e.preventDefault(); exec() }
  } // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<


  // mouse/drag/drop interactions >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  function contextMenu (e) { 
    e.preventDefault()
    var element = e.target
    var id = element.id
    var parent = element.parentElement // ?
    var cl = element.className
    var tag = element.tagName
    console.log(id, parent, cl)
  }

  function fileUpload (e) {
    e.preventDefault()
    var files = []
    var droppedFiles = e.dataTransfer.files
    var msg = R(messages, '<li id="{{id}}">{{body}}</li>')

    for (var i=0; i<droppedFiles.length; i++) {
      var f = droppedFiles[i]
      if (f.type.match('image.*')) files.push(f)
    }

    asyncMap(files, function handleFile (file, next) {
      var fileName = file.name
      var cleanName = file.name.replace(/ /g,'_')
      var id = utils.UID(cleanName)

      var formData = new FormData()
      formData.append('file', cleanName)

      var xhr = new XMLHttpRequest()
      xhr.upload.addEventListener('progress', function (e) { 
        var percent = (e.loaded / e.total)*100
        msg.draw({id: id, body: 'uploading ' + cleanName +' : '+ percent})
        if (e.loaded === e.total) setTimeout(function () {msg.erase()}, 1000)
      }, false) 
      xhr.open('post', '/upload', true) 

      var reader = new FileReader()

      reader.addEventListener('load', function (data) {
        formData.append('id', id)
        formData.append('blob', data.target.result)
        xhr.send(formData)
        next()
      }, false)

      reader.readAsDataURL(file)
    }, function End () {
      console.dir(droppedFiles)
    })
  } // <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<


  return s

}
