// omni/index.js
var through = require('through2') // fix this thing up!

module.exports = function Actions (C) {
  var omnibox
  var menu
  var search

  var s = through.obj(function Write (chunk, enc, next) { 
    console.log(chunk)
    next()
  })

  if (localStorage.library && localStorage.library.length > 0) {
    var library = JSON.parse(localStorage.library)
  } else library = null // request library

  window.addEventListener('contextmenu', ctx, false)

  function ctx (e) {
    e.preventDefault()
    var element = e.target
    var id = element.id
    var parent = element.parentElement // ?
    var cl = element.className
    var tag = element.tagName
    console.log(element, parent, parent.id, tag, cl)
  }
 
  // keybindings!
  window.addEventListener('keydown', function toggleOmni (e) {
    console.log(e)
    if (e.keyCode === 9) {
      e.preventDefault();
      (C.live === false) ? C.draw() : C.erase()
      return false
    } else if (e.keyCode === 38) {
      e.preventDefault()
      search.focus()
      return false
    } else if (e.keyCode === 37) {
      omni.prev()
    } else if (e.keyCode === 39) {
      omni.next()
    } else if (omni.live === true && e.keyCode === 13) {
      omni.go()
    }
  }, false) 


  function updateSearch (e) { // search materials --> send res to omni
    e.preventDefault()
    omnibox.innerHTML = ''
    var str = search.value
    var res = find(library, str) 

    if (res) // preview res
      omnibox.innerHTML = JSON.stringify(res) // package selected

    if (e.keyCode === 40 || e.keyCode === 39 || e.keyCode === 13) {
      console.log(e.keyCode)
      e.preventDefault()
      search.blur()
      if (res) omni._ = res
      omni.live = true
    }
  }

  var omni = {
    _ : null,
    live : false,
    next : function () { // modify cmds
      console.log('next!')
    },
    prev : function () {
      console.log('prev!')
    },
    go : function () {
      if (typeof this._ === 'object' && this._.nimoy) {
        var pkg = this._
        s.push({key:'draw', value: pkg})
      }
    }
  }
    
  // logic for processing omni results

  C.unbind = function () {
    search.removeEventListener('keyup', updateSearch)
  }

  C.bind = function (domNode) {
    omnibox = domNode.querySelector('.omnibox')
    search = domNode.querySelector('#search')
    search.addEventListener('keyup', updateSearch, false)
    search.focus()
  }

  return s
}

function find (haystack, needle) {
  for (hay in haystack) {
    if (needle !== '' && hay.match(needle)) return haystack[hay]
  }
}
